File: services/acceptance/src/test/java/robhopkins/iam/IAMRequests.java
package robhopkins.iam;

import io.restassured.RestAssured;
import io.restassured.http.Header;
import io.restassured.response.Response;
import io.restassured.specification.RequestSpecification;

import static org.assertj.core.api.Assertions.fail;

public final class IAMRequests {
    public static IAMRequests newInstance(final int port) {
        return new IAMRequests(String.format("http://localhost:%d", port));
    }

    private IAMRequests(final String baseUri) {
        RestAssured.baseURI = baseUri;
    }

    public void testReady() {
        final RequestSpecification request = RestAssured.given()
            .header(origin());
        // TODO: make me a bit cleaner.
        int count = 1;
        do {
            final int status = executeReadyQuery(request);
            if (status != 200) {
                count++;
                try {
                    Thread.sleep(100);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            } else {
                return;
            }
        } while (count < 11);
        fail("Service is not ready.");
    }

    private int executeReadyQuery(final RequestSpecification request) {
        return request.get("/canary").getStatusCode();
    }

    Response signin(final String username, final String secret) {
        return RestAssured.given()
            .body(signinForm(username, secret))
            .header(origin())
            .header("Content-Type", "application/x-www-form-urlencoded")
            .post("/iam/signin");
    }

    private Object signinForm(final String username, final String secret) {
        return String.format(
            "username=%s&secret=%s",
            username,
            secret
        );
    }

    Response signout(final String token) {
        return RestAssured.given()
            .header(origin())
            .header("Authorization", "Bearer " + token)
            .post("/iam/signout");
    }

    Response token(final String token) {
        return RestAssured.given()
            .header(origin())
            .header("Authorization", "Bearer " + token)
            .post("/iam/token");
    }

    private Header origin() {
        return new Header("Origin", "http://localhost");
    }

}

File: wc-ui/src/libs/IAMRequests.js
import { Token } from "./Token"

function createForm(username, password) {
  return `username=${username}&secret=${password}`;
  }

export function requestSignin(username, password, success = () => {}, failure = () => {} ) {
  fetch("/iam/signin", {
    method: "POST",
    body: createForm(username, password),
    headers: new Headers({
      'Content-Type': 'application/x-www-form-urlencoded'
      })
    })
    .then(response => {
      if (response.status === 200) {
        return response.json();
      } else {
        failure();
      }
    }).then(data => {
      success(new Token(data.token));
    }).catch(e => {
      failure();
    })
}

function authHeader() {
  const auth = sessionStorage.getItem("token");
  return `Bearer ${auth}`;
}

export function requestSignout(success = () => {}) {
  fetch("/iam/signout", {
      method: "POST",
      headers: new Headers({
        'Accept': 'application/json',
        'Authorization': authHeader()
      })
    }).then(response => {
      success();
      sessionStorage.removeItem("token");
    })
    .catch(e => {
      success();
      sessionStorage.removeItem("token");
    });
}